server/chat.ts

```diff
@@ -1803,29 +1806,84 @@ export const Chat = new class {
 	 * @param user - the user that sent the message
 	 * @param connection - the connection the user sent the message from
 	 */
-	parse(message: string, room: Room | null | undefined, user: User, connection: Connection) {
+	async parse(message: string, room: Room | null | undefined, user: User, connection: Connection) {
+		Monitor.activeIp = connection.ip;
 		Chat.loadPlugins();
 
 		const initialRoomlogLength = room?.log.getLineCount();
 		const context = new CommandContext({message, room, user, connection});
 		const start = Date.now();
-		const result = context.parse();
-		if (typeof result?.then === 'function') {
-			void result.then(() => {
-				this.logSlowMessage(start, context);
-			});
-		} else {
-			this.logSlowMessage(start, context);
-		}
+		const result = await context.parse();
+		this.logSlowMessage(start, context);
 		if (room && room.log.getLineCount() !== initialRoomlogLength) {
 			room.messagesSent++;
 			for (const [handler, numMessages] of room.nthMessageHandlers) {
 				if (room.messagesSent % numMessages === 0) handler(room, message);
 			}
 		}
-
+		Monitor.activeIp = null;
 		return result;
 	}
+
+	async queue(message: string, room: Room | null, connection: Connection) {
+		const user = connection.user;
+		const now = Date.now();
+		const noThrottle = user.hasSysopAccess() || Config.nothrottle;
+
+		if (message.startsWith('/cmd userdetails') || message.startsWith('>> ') || noThrottle) {
+			// certain commands are exempt from the queue
+			await Chat.parse(message, room, user, connection);
+			if (noThrottle) return;
+			return false; // but end the loop here
+		}
+
+		const throttleDelay = user.trusted ? Users.THROTTLE_DELAY_TRUSTED : Users.THROTTLE_DELAY;
+
+		if (user.chatQueueTimeout) {
+			if (!user.chatQueue) user.chatQueue = []; // this should never happen
+			if (user.chatQueue.length >= THROTTLE_BUFFER_LIMIT - 1) {
+				connection.sendTo(
+					room,
+					`|raw|<strong class="message-throttle-notice">Your message was not sent because you've been typing too quickly.</strong>`
+				);
+				return false;
+			} else {
+				user.chatQueue.push([message, room ? room.roomid : '', connection]);
+			}
+		} else if (now < user.lastChatMessage + throttleDelay) {
+			user.chatQueue = [[message, room ? room.roomid : '', connection]];
+			user.startChatQueue(throttleDelay - (now - user.lastChatMessage));
+		} else {
+			user.lastChatMessage = now;
+			await Chat.parse(message, room, user, connection);
+		}
+	}
+
+	async receive(message: string, room: Room | null, connection: Connection) {
+		const user = connection.user;
+		const multilineMessage = Chat.multiLinePattern.test(message);
+		if (multilineMessage) {
+			return Chat.queue(multilineMessage, room, connection);
+		}
+
+		const lines = message.split('\n');
+		if (!lines[lines.length - 1]) lines.pop();
+		// eslint-disable-next-line @typescript-eslint/prefer-optional-chain
```

Comment:
Why don't we use optional chaining here? It should be explained in the comment.

Reply:
Oh, I didn't write that part of the code - just moved it from user.chat. I assume it's because it used to be memory-heavy. CC @Zarel?

Reply:
ts-node completely fixes the perf issue with optional chaining; we can use it freely now.